modulus/models/gnn_layers/distributed_graph.py

```diff
@@ -0,0 +1,357 @@
+                global_node_features,
+                self.num_src_nodes_in_each_partition,
+                dim=0,
+                src=0,
+                group=self.process_group,
+            )
+
+        return global_node_features[self.partitioned_src_node_ids_to_global, :].to(
+            device=self.device_id
+        )
+
+    def get_src_node_features_in_local_graph(
+        self, partitioned_src_node_features: torch.Tensor
+    ) -> torch.Tensor:
+        # main primitive to gather all necessary src features
+        # which are required for a csc-based message passing step
+        return indexed_all_to_all_v(
+            partitioned_src_node_features,
+            indices=self.scatter_indices,
+            sizes=self.sizes,
+            use_fp32=True,
+            dim=0,
+            group=self.process_group,
+        )
+
+    def get_dst_node_features_in_partition(
+        self,
+        global_node_features: torch.Tensor,
+        scatter_features: bool = False,
+    ) -> torch.Tensor:
+        # if global features only on local rank 0 also scatter, split them
+        # according to the partition and scatter them to other ranks
+        if scatter_features:
+            return scatter_v(
+                global_node_features,
+                self.num_dst_nodes_in_each_partition,
+                dim=0,
+                src=0,
+                group=self.process_group,
+            )
+
+        return global_node_features[self.partitioned_dst_node_ids_to_global, :].to(
+            device=self.device_id
+        )
+
+    def get_dst_node_features_in_local_graph(
+        self,
+        partitioned_dst_node_features: torch.Tensor,
+    ) -> torch.Tensor:
+        # current partitioning scheme assumes that
+        # local graph is built from partitioned IDs
+        return partitioned_dst_node_features
+
+    def get_edge_features_in_partition(
+        self,
+        global_edge_features: torch.Tensor,
+        scatter_features: bool = False,
+    ) -> torch.Tensor:
+        # if global features only on local rank 0 also scatter, split them
+        # according to the partition and scatter them to other ranks
+        if scatter_features:
+            return scatter_v(
+                global_edge_features,
+                self.num_indices_in_each_partition,
+                dim=0,
+                src=0,
+                process_group=self.process_group,
+            )
+        return global_edge_features[self.partitioned_indices_to_global, :].to(
+            device=self.device_id
+        )
+
+    def get_edge_features_in_local_grpah(
+        self, partitioned_edge_features: torch.Tensor
+    ) -> torch.Tensor:
+        # current partitioning scheme assumes that
+        # local graph is built from partitioned IDs
+        return partitioned_edge_features
+
+    def get_global_src_node_features(
+        self,
+        partitioned_node_features: torch.Tensor,
+        get_on_all_ranks: bool = True,
```

Comment:
In relation to the comment above, I think it could be more elegant to add a more general "source-device" or a "destination device". This way, the user has more control on which device certain things should live - potentially even with CPU support.